_posts/2017-02-13-context-in-react.md

```diff
@@ -0,0 +1,249 @@
+intro: Today we'll dive into context in ReactJS, a feature that's often misunderstood and not used correctly by developers.
+githubPath: 2017-02-13-context-in-react
+---
+
+There is a lot of confusion amongst React developers on what context is, and why it exists. It's also a feature that's been hidden in the React documentation in the past and, [although it is now documented on the React site](https://facebook.github.io/react/docs/context.html) I thought a post on its usage and when to use it would be of use.
+
+The short answer is that you should __very rarely, if ever__ use context in your own React components. However, if you're writing a library of components, it can come in useful, and we'll discuss why this is later.
+
+## What is context in React, and how does it work?
+
+In React the primary mechanism for communication between your components is through properties, or `props`, for short. Parent components can pass properties down to their children:
+
+```js
+const SomeParentComponent = () => {
+  const foo = 2
+  return (
+    <SomeChildComponent foo={foo} />
+  )
+}
+```
+
+Here, the parent component `SomeParentComponent` passes the prop `foo` through to its child, `SomeChildComponent`.
+
+> Here, a _child component_ is a component that another component renders. A _parent component_ is a component that directly renders another.
+
+If a child component wants to communicate back to its parent, it can do so through props, most commonly by its parent providing a _callback property_ that the child can call when some event happens:
+
+```js
+const SomeParentComponent = () => {
+  const letMeKnowAboutSomeThing = () => console.log('something happened!')
+
+  return (
+    <SomeChildComponent letMeKnowAboutSomeThing={letMeKnowAboutSomeThing} />
+  )
+}
+
+const SomeChildComponent = props => {
+  const onClick = e => {
+    e.preventDefault()
+    props.letMeKnowAboutSomeThing()
+  }
+
+  return <a onClick={onClick}>Click me!</a>
+}
+```
+
+The key thing about this communication is that it's _explicit_. Looking at the code above, you know how the components are communicating, where the `letMeKnowAboutSomeThing` function comes from, who calls it, and which two components are in communication. [You can see this in action on CodePen](http://codepen.io/jackfranklin/pen/vgvYOa?editors=0011).
+
+This property of React, its explicitness of data passing between components, is one of its best features. React is very explicit as a rule, and this is in my experience leads to clearer code that's much easier to maintain and debug when something goes wrong. You simply have to follow the path of props to find the problem.
+
+As you might have guessed, context is a way to break this explicitness. When a component defines some data onto its _context_, any of its descendants can access that data. That means any child further down in the component tree can access data from it, without being passed it as a property. Let's take a look at context in action.
```

Comment:
I like to think of `context` as a _props portal_ between a `Parent` and a `Child`. hat do you think about using this wording to describe what `context` is?